lib/api/api.js

```diff
@@ -26,11 +29,50 @@ auth.setAuthHandler(vault);
 
 const api = {
     callApiMethod(apiMethod, request, log, callback, locationConstraint) {
+        if (apiMethod === 'objectCopy') {
+            let source =
+                querystring.unescape(request.headers['x-amz-copy-source']);
+            // If client sends the source bucket/object with a leading /,
+            // remove it
+            if (source[0] === '/') {
+                source = source.slice(1);
+            }
+            const slashSeparator = source.indexOf('/');
+            if (slashSeparator === -1) {
+                return callback(errors.InvalidArgument);
+            }
+            // Pull the source bucket and source object separated by /
+            const sourceBucket = source.slice(0, slashSeparator);
+            const sourceObject = source.slice(slashSeparator + 1);
+            const getRequestContext = new RequestContext(request.headers,
+                request.query, sourceBucket, sourceObject,
+                request.socket.remoteAddress, request.connection.encrypted,
+                'objectGet', 's3', locationConstraint);
+
+            return auth.doAuth(request, log, err => {
```

Comment:
An extra call for fail requests is not desired.  I don't think it would be much of a benefit for successful and there is already so much going on that I prefer to keep it separate.